pkg/schema/definitions/handler.go

```diff
@@ -0,0 +1,183 @@
+// cached models found during the last refresh as part of this process
+func (s *SchemaDefinitionHandler) ByIDHandler(request *types.APIRequest) (types.APIObject, error) {
+	// pseudo-access check, designed to make sure that users have access to the schema for the definition that they
+	// are accessing.
+	requestSchema := request.Schemas.LookupSchema(request.Name)
+	if requestSchema == nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.NotFound, "no such schema")
+	}
+
+	if s.needsRefresh() {
+		err := s.refresh()
+		if err != nil {
+			logrus.Errorf("error refreshing schemas %s", err.Error())
+			return types.APIObject{}, apierror.NewAPIError(internalServerErrorCode, "error refreshing schemas")
+		}
+	}
+
+	// lock only in read-mode so that we don't read while refresh writes. Only use a read-lock - using a write lock
+	// would make this endpoint only usable by one caller at a time
+	s.RLock()
+	defer s.RUnlock()
+
+	if s.models == nil {
+		return types.APIObject{}, apierror.NewAPIError(notRefreshedErrorCode, "schema definitions not yet refreshed")
+	}
+	models := *s.models
+	modelName, ok := s.schemaToModel[requestSchema.ID]
+	if !ok {
+		return types.APIObject{}, apierror.NewAPIError(notRefreshedErrorCode, "no model found for schema, try again after refresh")
+	}
+	model := models.LookupModel(modelName)
+	protoKind, ok := model.(*proto.Kind)
+	if !ok {
+		errorMsg := fmt.Sprintf("model for %s was type %T, not a proto.Kind", modelName, model)
+		return types.APIObject{}, apierror.NewAPIError(internalServerErrorCode, errorMsg)
+	}
+	definitions := map[string]definition{}
+	visitor := schemaFieldVisitor{
+		definitions: definitions,
+		models:      models,
+	}
+	protoKind.Accept(&visitor)
+
+	return types.APIObject{
+		ID:   request.Name,
+		Type: "schemaDefinition",
+		Object: schemaDefinition{
+			DefinitionType: modelName,
+			Definitions:    definitions,
+		},
+	}, nil
+}
+
+// needsRefresh readLocks and checks if the cache needs to be refreshed
+func (s *SchemaDefinitionHandler) needsRefresh() bool {
+	s.RLock()
+	defer s.RUnlock()
+	if s.lastRefresh.IsZero() {
+		return true
+	}
+	return s.lastRefresh.Add(s.refreshStale).Before(time.Now())
+}
+
+// refresh writeLocks and updates the cache with new schemaDefinitions. Will result in a call to OpenAPISchemas
+func (s *SchemaDefinitionHandler) refresh() error {
+	s.Lock()
+	defer s.Unlock()
+	openapi, err := s.client.OpenAPISchema()
+	if err != nil {
+		return fmt.Errorf("unable to fetch openapi definition: %w", err)
+	}
+	models, err := proto.NewOpenAPIData(openapi)
+	if err != nil {
+		return fmt.Errorf("unable to parse openapi definition into models: %w", err)
+	}
+	s.models = &models
+	nameIndex, err := s.indexSchemaNames(models)
+	if err != nil {
+		return fmt.Errorf("unable to index schema name to model name: %w", err)
+	}
+	s.schemaToModel = nameIndex
+	s.lastRefresh = time.Now()
+	return nil
+}
```

Comment:
nit: 
Since needsRefresh is always followed by a refresh, you could have a function that is refreshIfNeeded that performs the checks and then calls the refresh for the user. This is just a stylistic suggestion, and you can resolve it if you disagree.

Reply:
I'm going to leave this as-is for the time being. There's other reasons that we could want to refresh (for example, currently we don't refresh if we have a schema but no model), so I think that they make sense as separate functions. I don't feel too strongly about it since right now it doesn't function as is, so I'll leave this thread open in case other reviewers feel strongly in this area.